runner/master_log.py

```diff
@@ -0,0 +1,172 @@
+        if test_id:
+            event = LogEvent(EventType.SEND_MESSAGE, timestamp, test_id,  {"message": message, "is_question": is_question})
+        else:
+            event = LogEvent(EventType.SEND_FILL, timestamp, data={"message": message, "is_question": False})
+
+        self.add_event(event)
+
+    def add_response_message(self, message: str, timestamp: datetime, test_id: Optional[str] = None, is_question: Optional[bool] = False):
+        if test_id:
+            event = LogEvent(EventType.RESPONSE_MESSAGE, timestamp, test_id,  {"message": message, "is_question": is_question})
+        else:
+            event = LogEvent(EventType.RESPONSE_FILL, timestamp, data={"message": message, "is_question": False})
+
+        self.add_event(event)
+
+    def add_wait_event(self, test_id: str, timestamp: datetime, tokens=0, time=0):
+        event = LogEvent(EventType.WAIT, timestamp=timestamp, test_id=test_id, data={"tokens":tokens, "time":time})
+        self.add_event(event)
+
+    def begin_test(self, test_id, timestamp):
+        event = LogEvent(EventType.BEGIN, timestamp=timestamp, test_id=test_id)
+        self.add_event(event)
+
+    def end_test(self, test_id: str, timestamp: datetime):
+        event = LogEvent(EventType.END, timestamp=timestamp, test_id=test_id)
+        self.add_event(event)
+
+    def add_event(self, event: LogEvent):
+        self.log.append(event)
+
+    def human_readable_full_log(self, test_id: str, message: str) -> List[str]:
+        # Collate all the messages from the index point
+        messages = []
+        index = self.find_message(test_id, message)
+
+        for event in self.log[index:]:
+            if event.type in [EventType.SEND_MESSAGE, EventType.RESPONSE_MESSAGE, EventType.SEND_FILL, EventType.RESPONSE_FILL]:
+                sender = "Test" if event.type == EventType.SEND_MESSAGE else "Agent"
+                messages.append(f"{sender} ({event.timestamp}): {event.data['message']}")
+            elif event.type == EventType.WAIT:
+                messages.append(f"SYSTEM ({event.timestamp}): Test '{event.test_id}' WAITING for {event.data['tokens']} tokens until {event.data['time']}")
+            elif event.type == EventType.BEGIN:
+                messages.append(f"SYSTEM ({event.timestamp}): Test '{event.test_id}' BEGINS")
+            elif event.type == EventType.END:
+                messages.append(f"SYSTEM ({event.timestamp}): Test '{event.test_id}' ENDS")
+            else:
+                messages.append(f"SYSTEM ({event.timestamp}):  UNKNOWN EVENT")
+
+        return messages
+
+    def find_message(self, test_id: str, message: str) -> int:
+        index = 0
+        # Find the index of the sent message
+        for event in self.log:
+            if event.test_id == test_id and event.type == EventType.SEND_MESSAGE and event.data["message"] == message:
+                break
+            index += 1
+
+        return index
```

Comment:
This method will always return a valid index, even if the message is not found. The standard with `find` is to return `-1` when that happens, but I think an exception is more useful in this case.